File: fixtures/scale/scala/large/src/cryptobench/ciphers/twosquare/TwoSquareCipher.scala
package cryptobench.ciphers.twosquare

import cryptobench.ciphers.playfair.PlayfairDigraphs
import cryptobench.core.Cipher
import cryptobench.util.Alphabet

/** Digraphs looked up across two keyword squares stacked vertically. */
final class TwoSquareCipher(key: TwoSquareKey) extends Cipher {
  private val top = key.topSquare
  private val bottom = key.bottomSquare

  override def name: String = "twosquare"

  override def encrypt(plaintext: String): String = swap(PlayfairDigraphs.split(Alphabet.clean(plaintext)))

  override def decrypt(ciphertext: String): String = swap(PlayfairDigraphs.split(Alphabet.clean(ciphertext)))

  private def swap(pairs: String): String = {
    val sb = new StringBuilder
    var i = 0
    while (i + 1 < pairs.length) {
      val a = top.indexOf(pairs.charAt(i))
      val b = bottom.indexOf(pairs.charAt(i + 1))
      sb.append(top.charAt(a / 5 * 5 + b % 5)).append(bottom.charAt(b / 5 * 5 + a % 5))
      i += 2
    }
    sb.toString
  }
}

File: fixtures/scale/dart/large/lib/ciphers/twosquare/twosquare_cipher.dart
import '../../core/cipher.dart';
import '../../util/alphabet.dart';
import '../playfair/playfair_digraphs.dart';
import 'twosquare_key.dart';

/// Digraphs looked up across two keyword squares stacked vertically.
class TwoSquareCipher implements Cipher {
  final String _top;
  final String _bottom;

  TwoSquareCipher(TwoSquareKey key)
      : _top = key.topSquare(),
        _bottom = key.bottomSquare();

  @override
  String name() => 'twosquare';

  @override
  String encrypt(String plaintext) => _swap(splitDigraphs(cleanAlphabet(plaintext)));

  @override
  String decrypt(String ciphertext) => _swap(splitDigraphs(cleanAlphabet(ciphertext)));

  String _swap(String pairs) {
    final sb = StringBuffer();
    var i = 0;
    while (i + 1 < pairs.length) {
      final a = _top.indexOf(pairs[i]);
      final b = _bottom.indexOf(pairs[i + 1]);
      sb.write(_top[a ~/ 5 * 5 + b % 5]);
      sb.write(_bottom[b ~/ 5 * 5 + a % 5]);
      i += 2;
    }
    return sb.toString();
  }
}
